Add target groups for improved interface

alairock opened the pull request:
This pull request introduces a new "Target Groups" feature to the Warpgate application, allowing targets to be organized into groups with associated metadata (name, description, color). It adds database support, API endpoints for managing target groups, and updates the targets to optionally reference a group. The HTTP protocol and web UI are also updated to expose and utilize this grouping functionality.

**Target Groups Feature**

* Added new `TargetGroup` entity, database migration, and ORM model, including fields for `id`, `name`, `description`, and `color`. [[1]](diffhunk://#diff-67a95652ef3f605b692272e63d2d51e22f14002a127119f4cba7113a30d2b0b9R1-R25) [[2]](diffhunk://#diff-175c343e2be9f946c88f9d90b1620c79403f0cca74afd7c4aa43f33e4dfda4c7R1-R102)
* Created REST API endpoints for listing, creating, retrieving, updating, and deleting target groups in `target_groups.rs`, and registered these endpoints in the main API module. [[1]](diffhunk://#diff-f9e2ce7e2efafd2fe1cb52ed5fc182854cbf148ec2a4f171405b3f787fb87110R1-R208) [[2]](diffhunk://#diff-ad95ed5547cb582058361b260d52dd0503ce7b1b0ae21974a238792189346b11R19) [[3]](diffhunk://#diff-ad95ed5547cb582058361b260d52dd0503ce7b1b0ae21974a238792189346b11R36)

**Targets Integration**

* Extended target models and database schema to include an optional `group_id` field, and added the relation between targets and target groups. [[1]](diffhunk://#diff-a5b6f836b60eb623474d99de60f74a3c8410b2b347dac087f0e5913b672b11e1R55) [[2]](diffhunk://#diff-a5b6f836b60eb623474d99de60f74a3c8410b2b347dac087f0e5913b672b11e1L68-R79) [[3]](diffhunk://#diff-a5b6f836b60eb623474d99de60f74a3c8410b2b347dac087f0e5913b672b11e1R95) [[4]](diffhunk://#diff-bb051c46669335f59bf38e0bfb2ad73458823bcaf5fc052b8c75fa1b06da86d3R29) [[5]](diffhunk://#diff-84d2e0037cb4184806d5bebb10807f5254713558c79a88cc66553a00ae3374eeR142) [[6]](diffhunk://#diff-b685c67434cc98c054511b9090996823ac710742328a0c136958d9282d2e741fR122) [[7]](diffhunk://#diff-175c343e2be9f946c88f9d90b1620c79403f0cca74afd7c4aa43f33e4dfda4c7R1-R102)
* Updated targets API endpoints to support filtering by group, setting group on create/update, and returning group information in responses. [[1]](diffhunk://#diff-bb051c46669335f59bf38e0bfb2ad73458823bcaf5fc052b8c75fa1b06da86d3R59) [[2]](diffhunk://#diff-bb051c46669335f59bf38e0bfb2ad73458823bcaf5fc052b8c75fa1b06da86d3R71-R74) [[3]](diffhunk://#diff-bb051c46669335f59bf38e0bfb2ad73458823bcaf5fc052b8c75fa1b06da86d3R117) [[4]](diffhunk://#diff-bb051c46669335f59bf38e0bfb2ad73458823bcaf5fc052b8c75fa1b06da86d3R214) [[5]](diffhunk://#diff-0535d25b685b2bfa4cd400006aef2c054ae137396bfa25e1417e662598c75cefR6-R10) [[6]](diffhunk://#diff-0535d25b685b2bfa4cd400006aef2c054ae137396bfa25e1417e662598c75cefR23-R25) [[7]](diffhunk://#diff-0535d25b685b2bfa4cd400006aef2c054ae137396bfa25e1417e662598c75cefL83-R143)

**Database Migration**

* Added migration `m00020_target_groups` to create the `target_groups` table and add the `group_id` column to the `targets` table. [[1]](diffhunk://#diff-ae45949a134d49f12101e8550287a30919f67e7efb84f5fafd11457996b6af0aR24) [[2]](diffhunk://#diff-ae45949a134d49f12101e8550287a30919f67e7efb84f5fafd11457996b6af0aR51) [[3]](diffhunk://#diff-175c343e2be9f946c88f9d90b1620c79403f0cca74afd7c4aa43f33e4dfda4c7R1-R102)

**Web UI Integration**

* Registered new routes for target group management in the admin Svelte application.

**Other**

* Registered the new entity and migration in the appropriate modules. [[1]](diffhunk://#diff-412ba1c63d322d1af11146676d0665c52e3f60ac2fb48a066bc4cd056fa16803R15) [[2]](diffhunk://#diff-ae45949a134d49f12101e8550287a30919f67e7efb84f5fafd11457996b6af0aR24)

These changes collectively enable users to manage target groups, associate targets with groups, and view group information throughout the API and UI.

<img width="1333" height="772" alt="CleanShot 2025-09-30 at 18 35 53" src="https://github.com/user-attachments/assets/d5329c0d-b569-4f1c-9052-82ad8098a802" />


<img width="1375" height="775" alt="CleanShot 2025-09-30 at 18 35 47" src="https://github.com/user-attachments/assets/a68fb42b-607c-4269-8873-b7ea56c38024" />


<img width="928" height="992" alt="CleanShot 2025-09-30 at 18 35 34" src="https://github.com/user-attachments/assets/3b5f9daa-ffe5-42ab-bc6f-7ea83fd605b1" />

Christian-Little-Tava commented:
This is a cool idea, we have a lot of targets as well so far and we were wondering on the best ways to organize them to make it easier for the users

Eugeny commented on warpgate-protocol-http/src/api/targets_list.rs:
```diff
@@ -19,6 +20,9 @@ pub struct TargetSnapshot {
     pub description: String,
     pub kind: Target::TargetKind,
     pub external_host: Option<String>,
+    pub group_id: Option<uuid::Uuid>,
```
Should be a nested `{id, name, color}` object

Eugeny commented on warpgate-db-entities/src/TargetGroup.rs:
```diff
@@ -0,0 +1,25 @@
+    pub name: String,
+    #[sea_orm(column_type = "Text")]
+    pub description: String,
+    pub color: Option<String>, // Hex color code for UI display
```
For cleaner UI, please restrict the color to existing bootstrap theme-colors (`primary`, `secondary`, `danger` etc) - this lets you plug into existing components (e.g. `badge`) and avoids the need to manually generate contrasting font colors etc

alairock commented:
Realized I had hard-coded hex numbers for colors, and not used the name of the style itself. Fixed that. 

I also updated the target groups edit page to use a visual selection, instead of just having a drop down with the bootstrap names. That way they can see the color they are picking at the time of picking it. 

<img width="1356" height="835" alt="CleanShot 2025-11-17 at 09 42 03" src="https://github.com/user-attachments/assets/9f86654c-5dd4-4d42-adb0-9029716de416" />

Eugeny commented:
Thank you!

alairock commented:
> Looks like an awesome feature! Could it do permissions at the group level? It might be out of scope since it's primarily an interface improvement, but it would be nice to give a role access to a whole group.

@Eugeny I'm not sure what you mean? Would be happy to look into it to see if I can add that feature :)
